python/thunder/imgprocessing/transformation.py

```diff
@@ -8,6 +9,168 @@ def apply(self, im):
         raise NotImplementedError
 
 
+class DifferentiableTransformation(Transformation):
+    """
+    Differentiable transformations must have methods to compute the Jacobian and update parameters. These are used by
+    iterative alignment techniques like Lucas-Kanade and RASL.
+    """
+    def jacobian(self, imageGradients, imageGrid):
+        """Compute gradient of transformation output with respect to parameters.
+
+        Parameters
+        ----------
+        imageGradients : list of volumes, one for each dimension
+            gradient of volume in each dimension
+        imageGrid : list of arrays, one for each dimension
+            each array contains the coordinates along that dimension for each voxel
+
+        Returns
+        -------
+        ordered list of arrays containing the jacobian for each parameter
+        """
+        raise NotImplementedError
+
+    def updateParams(self, deltaParams):
+        """Update each parameter in the transformation.
+
+        Parameters
+        ----------
+        deltaParams : array, shape (nparams,)
+            ordered changes in parameters to be applied. The ordering here will match
+            the ordering of the parameters returned by jacobian.
+        """
+        raise NotImplementedError
+
+
+class AffineTransformation(DifferentiableTransformation):
```

Comment:
Maybe don't use `Transformation` in the names for all these? Seems redundent given the class hierarchy, and is inconsistent with the others (like `Displacement`).